lib/ui/window.dart

```diff
@@ -286,31 +543,197 @@ class Locale {
+        if (entry.value.isNotEmpty)
+          tOut.write('-${entry.value}');
+      } else {
+        throw UnimplementedError(
+            'Extension not supported/recognised: $entry.');
+      }
+    }
+    if (tAttr != null || tOut.isNotEmpty) {
+      result.write('-t');
+      if (tAttr != null)
+        result.write('-$tAttr');
+      result.write(tOut.toString());
+    }
+    if (uAttr != null || uOut.isNotEmpty) {
+      result.write('-u');
+      if (uAttr != null && uAttr.isNotEmpty)
+        result.write('-$uAttr');
+      result.write(uOut.toString());
+    }
+    if (resultVWYZX.isNotEmpty)
+      result.write(resultVWYZX.toString());
+    if (extensions.containsKey('x')) {
+      result.write('-x');
+      if (extensions['x'].isNotEmpty) {
+        result.write('-${extensions["x"]}');
+      }
+    }
+    return result.toString();
+  }
+
+  // Unicode Language Identifier subtags
+  // TODO/WIP: because we lowercase Locale Identifiers before parsing, typical
+  // use of these regexps don't actually need to atch capitals too.
+  static final RegExp _reSingleton = RegExp(r'^[a-zA-Z]$');
+
+  // (https://www.unicode.org/reports/tr35/#Unicode_language_identifier).
+  static final RegExp _reLanguage = RegExp(r'^[a-zA-Z]{2,3}$|^[a-zA-Z]{5,8}$');
+  static final RegExp _reScript = RegExp(r'^[a-zA-Z]{4}$');
+  static final RegExp _reRegion = RegExp(r'^[a-zA-Z]{2}$|^[0-9]{3}$');
+  static final RegExp _reVariant = RegExp(r'^[a-zA-Z0-9]{5,8}$|^[0-9][a-zA-Z0-9]{3}$');
+  static final RegExp _reSep = RegExp(r'[-_]');
+
+  // Covers all subtags possible in Unicode Locale Identifiers, used for
+  // pu_extensions.
+  static final RegExp _reAllSubtags = RegExp(r'^[a-zA-Z0-9]{1,8}$');
+  // Covers all subtags within a particular extension, used for other_extensions.
+  static final RegExp _reOtherSubtags = RegExp(r'^[a-zA-Z0-9]{2,8}$');
+  // Covers "attribute" and "type" from unicode_locale_extensions, and "tvalue" in
+  // transformed_extensions.
+  // (https://www.unicode.org/reports/tr35/#Unicode_locale_identifier).
+  static final RegExp _reValueSubtags = RegExp(r'^[a-zA-Z0-9]{3,8}$');
+
+  static final RegExp _reKey = RegExp('^[a-zA-Z0-9][a-zA-Z]\$');
+  static final RegExp _reTkey = RegExp('^[a-zA-Z][0-9]\$');
+
   @override
   bool operator ==(dynamic other) {
     if (identical(this, other))
       return true;
     if (other is! Locale)
       return false;
     final Locale typedOther = other;
-    return languageCode == typedOther.languageCode
+
+    // TODO: improve efficiency of this?
+    // Comparing Sets and Maps requires reimplementing functionality in
+    // package:collection, comparing canonical string is simple.
+    return toLanguageTag() == typedOther.toLanguageTag()
+        // toLanguageTag() cannot represent zero-length string as country-code,
+        // but we need to distinguish it for backward compatibility reasons.
         && countryCode == typedOther.countryCode;
   }
```

Comment:
this `==` is far too expensive now, IMHO.

Reply:
hashCode even worse. Adjusting...

Reply:
Fixed - how does it look to you now? Typical case now has an empty-list-join, and function call that does a check and returns ''.

Reply:
still seem slower than just doing a list comparison? Comparing the two variants lists and the two extensions sets is trivial, as far as I can tell.

Reply:
List comparison is shallow, we need deep:
  `assert([1, 2, 3] != [1, 2, 3]);  // Assertion passes.`
So there is `package:collection` that we can't use, because we're in a `dart:*` SDK library.
https://pub.dartlang.org/documentation/collection/latest/collection/MapEquality-class.html

I'm happy to try optimising variants and extensions comparisons too by manually implementing deep comparisons. More code complexity for rare tail performance optimisation? (Will mark this conversation resolved if I've done that.)

Reply:
List comparison isn't even shallow (one level deep), it's just identity comparison of the list object.